### DISCUSSION TITLE
Clarification on the MSBuild property RestorePackagesConfig

### DISCUSSION THREAD

**Discussion by Bartleby2718** [NuGet/Home#13534] [Q&A] [answered]
### Goal
I want to be able to (restore and) build both SDK-style and non-SDK C# projects using a single MSBuild command.

### Thought Process
1. `MSBuild.exe MySolution.sln ... -restore` works for SDK-style projects that use `PackageReference`s, but I also want to support non-SDK-style projects that use `packages.config`.
2. Based on https://learn.microsoft.com/en-us/nuget/consume-packages/package-restore, it seems like `-p:RestorePackagesConfig=true` is all I need. I confirmed that this works for non-SDK-style projects that use `packages.config`.

### Questions
1. **Does `MSBuild.exe MySolution.sln ... -restore -p:RestorePackagesConfig=true` also work for SDK-style projects that use `PackageReference`s?** It worked for one such solution, but I'm not sure if I got lucky or this is the intended behavior. FWIW https://stackoverflow.com/a/73575473 says it works.
2. Not important if the answer is yes to the first question, but if possible, I want to move `RestorePackagesConfig` to a file. Therefore, I created `Directory.Build.props` next to the solution file:
```xml
<Project>
    <PropertyGroup>
      <RestorePackagesConfig>true</RestorePackagesConfig>
    </PropertyGroup>
</Project>
```
However, a bunch of `<Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">`s are still mad at me because the package restore isn't being done. (This target is mentioned in #3320, but I feel like this target is doing its job in this case.) For some reason, I get
>   Determining projects to restore...
>  All projects are up-to-date for restore.

even though the projects are not.

### Notes
- In https://github.com/NuGet/Home/discussions/12554 and https://github.com/NuGet/Home/discussions/12982, @zivkan does "strongly recommend you stop using nuget.exe, and use [msbuild -t:restore -p:RestorePackagesConfig=true](https://learn.microsoft.com/en-us/nuget/reference/msbuild-targets#restoring-packagereference-and-packagesconfig-projects-with-msbuild) instead" but it's unclear whether it'll work for SDK-style projects that use `PackagReference`s.

**Comment by zivkan** [comment, marked as answer]
1. yes, msbuild works on SDK style projects. I think of Visual Studio as a GUI on top of MSBuild, so in most (but unfortunately not all) cases, if it works in VS, it'll probably work on the CLI with msbuild.

2. msbuild restore of packages.config doesn't work on a per-project setting. It needs to be passed into Nuget's RestoreTask. This is why I specified the `-p:RestorePackagesConfig=true` on the command line, which personally I recommend in CI scripts as well, but that's just my personal preference. But in theory copying that `Directory.Build.props` example to `Directory.Solution.props` will work: https://learn.microsoft.com/en-us/visualstudio/msbuild/customize-solution-build?view=vs-2022

**Reply by Bartleby2718** [reply]
Thanks @zivkan for the response! A couple more questions:
1. [I didn't find anything about `RestorePackagesConfig` in `dotnet/msbuild`](https://github.com/search?q=repo%3Adotnet%2Fmsbuild%20RestorePackagesConfig&type=code), but I did see [a few occurrences of `RestorePackagesConfig` in `NuGet/NuGet.Client`](https://github.com/search?q=repo%3ANuGet/NuGet.Client%20RestorePackagesConfig&type=code). When `RestorePackagesConfig` is set, can you tell from the code whether MSBuild restores packages _only_ from `packages.config` or _also_ `packages.config`? I think that'll give me extra confidence.
2. Huh, I had tried both `Directory.Build.props` and `before.MySolution.sln.targets` and thought neither worked. (I didn't know about `Directory.Solution.props`.) Now I see that `Directory.Solution.props` and `before.MySolution.sln.targets` work, which matches what you said, and `Directory.Build.props` doesn't. At least I have a working solution, so this is great!

**Reply by zivkan** [reply]
It restores both packages.config and PackageReference. [Here's the code that does the pacakges.config restore](https://github.com/NuGet/NuGet.Client/blob/de9c314080687bfe4da4baf4b3d1a9d38b6ace3a/src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs#L185-L211), and immediately after that block is the PackageReference block. There's no `else` statement, or early `return` statement.
